### DISCUSSION TITLE
[React-Challenge] How we can Implement React Challengee

### DISCUSSION THREAD

**Discussion by Roshan-Horo** [reactplay/react-play#1196] [General]
## Objective

We have following points as an objective/goals, it's can add/update overtime

- We can create a challenge ( how we can create a challenge is not defined yet )
- people can take the challenge

## Technical Requirement

Minimum Requirement as of now, 

- Code Editor & it's preview
- Can Run Test on the code

## Initial layout design

We can Imagine rough design layout of the react challenge is like this where we have : 

![Screenshot 2023-06-28 201133](https://github.com/reactplay/react-play/assets/35597955/e7e1a5a5-9c18-4832-be39-632bd8c902b6)

- ` Challenge Description ` : All information related to challenge.
- ` Final Results ( image/screenshot ) ` : A demo about the final result.

We can collapse these two options ( challenge description & final results ) , so that we got more room for coding, which has

![Screenshot 2023-06-28 200812](https://github.com/reactplay/react-play/assets/35597955/513445eb-9d60-48a4-afc5-3e6d962d6bf7)

- ` Editor ( Coding & preview ) ` : A code editor and a preview section.
- ` Test Results ` : Show the results of the test cases.

## Different Ways to add Code Editor

We can differentiate code editors based on where it can run our code so we have

- ` Server code editor ` : these can run our code on their platform/server and gives client/embed for embedding on the site using sandbox link. e.g : cloud sandbox. ( we can run any language C/C++, java, javascript, any modern framework ( Next.js, Sveltkit ), etc.
![Screenshot (294)](https://github.com/reactplay/react-play/assets/35597955/4b59604f-d503-4238-ade9-e3436f3a7411)

- ` Browser code editor ` : these can run our code on Browser. we have lot more variation and also through new technology like web containers. These can be following types example

1.` Browser Sandbox `: This can run javascript/react on the browser through embedding and sandbox link ( which works for platforms like Medium, Reddit, Trello, and Notion ). [docs link]: https://codesandbox.io/docs/learn/sandboxes/embedding?tab=browser-sandboxes
![Screenshot (295)](https://github.com/reactplay/react-play/assets/35597955/987bc91f-0d90-4c44-a809-69d95ce9f5a4)

2. ` Sandpack from sandbox ` : This can run javascript/react on the browser through npm package i.e ` @codesandbox/sandpack-react`, and can show console, file explorer and other configs also. New react docs site uses this for it's challenge part , you can check it out [here]: https://react.dev/learn/reacting-to-input-with-state#challenges

### Simple sandpack example
![Screenshot (297)](https://github.com/reactplay/react-play/assets/35597955/eb38e78c-fb05-462d-a590-1eef055b11a2)

### Can be configured based on the needs
![Screenshot (300)](https://github.com/reactplay/react-play/assets/35597955/cc79d9c5-7e27-4f34-93bb-3949397d4383)

![Screenshot (302)](https://github.com/reactplay/react-play/assets/35597955/0862db1b-9234-4029-9acf-7e8484413824)

### It has an option for test result if you have test cases, but it's really slow and not working in my demo
![Screenshot (301)](https://github.com/reactplay/react-play/assets/35597955/3af9f1dd-dad5-41d1-8450-d2534dc246df)

3. ` Nodebox from sandbox and WebContainer from stackblitz ` : These can run node on the browser and support other npm dependencies out of the box . Still learning about this and how we can use it. [nodebox]: https://sandpack.codesandbox.io/docs/advanced-usage/nodebox , [webcontainer]: https://webcontainers.io/

## Problems/Stuck 

We need three things to be good - Code Editor, Preview and testing for the code. We get the Editor and Preview from the sandpack as 
it fits really well in our case but we didn't get that level of testing ( like unit testing through jest or react testing library ) in sandpack.
I'm checking other options to integrate better testing.

## Reference links

can checkout the demo app of all these above example using [github repo](https://github.com/Roshan-Horo/react_challenge_demo)

**Comment by debjit** [comment]
This is a custom codebase, and I am working on implementing this on my other projects, so this will be fun.
https://www.freecodecamp.org/news/build-and-deploy-a-leetcode-clone-with-react-next-js-typescript-tailwind-css-firebase/

**Reply by debjit** [reply]
We can use any opensource alternative for this one, like nhost, appwrite, superbase.
